src/Adapter/Product/Combination/Create/CombinationCreator.php

```diff
@@ -125,54 +152,76 @@ private function formatScalarValues(array $groupedAttributeIdsList): array
     /**
      * @param Product $product
      * @param Traversable $generatedCombinations
+     * @param ShopId[] $shopIds
      *
-     * @return CombinationId[]
-     *
-     * @throws CoreException
+     * @return CombinationId[] the ids of newly created combinations
      */
-    private function addCombinations(Product $product, Traversable $generatedCombinations): array
+    private function addCombinations(Product $product, Traversable $generatedCombinations, array $shopIds): array
     {
         $product->setAvailableDate();
         $productId = new ProductId((int) $product->id);
-        $alreadyHasCombinations = $hasDefault = (bool) $this->combinationRepository->findDefaultCombination($productId);
+        $hasCombinations = $this->productRepository->hasCombinations($productId);
+        $newCombinationIds = [];
+
+        foreach ($shopIds as $shopId) {
```

Comment:
I think you should "reverse" those to loops. First loop on the combinations to generate, and inside that loop loop through the shops This way you can optimize the creation of totally missing combinations but calling instantly:

```php
$newCombinationIds[] = $this->persistCombination($productId, $generatedCombination, $shopIds);
```

You'll need to have an array parameter for `persistCombination` and for `CombinationRepository::create`

Reply:
This way you only need to loop through the `$shopIds` after you checked if `$matchingCombinationId` already exists and the second part of the loop is only dedicated to shop association.